### ISSUE TITLE
sql: drop table and recreate fail

### ISSUE THREAD

**Issue by idsj** [cockroachdb/cockroach#7348]
Please follow the steps below to help us help you.
1. header
   I160621 11:06:31.940101 util/log/clog.go:998  [config] file created at: 2016/06/21 11:06:31
   I160621 11:06:31.940101 util/log/clog.go:998  [config] running on machine: longong-dsj-kv1
   I160621 11:06:31.940101 util/log/clog.go:998  [config] binary: CockroachDB beta-20160616 (linux amd64, built 2016/06/16 16:51:45, go1.6.2)
   I160621 11:06:31.940101 util/log/clog.go:998  [config] arguments: [./cockroach start --host=192.168.181.91 --insecure --join=192.168.181.92:26258]
2. Please describe the issue you observed:
- What did you do?
  e.g 
  root@192.168.181.91:26257> drop table test_serial;
  DROP TABLE
  root@192.168.181.91:26257> select count(*) from test_serial;
  pq: table is being deleted
  root@192.168.181.91:26257> show tables;
  +-------------+
  |    Table    |
  +-------------+
  | test_serial |
  +-------------+
- What did you expect to see?
  create a new table with same name
- What did you see instead?
  i can not create a new table with same name
  e.g 
  root@192.168.181.91:26257> create table test_serial(a serial primary key,b string(30),c bool);
  pq: table "test_serial" already exists

**Comment by andreimatei** [comment]
Are you running the `DROP` statement in a transaction? If so, DDL statements only take effect when you commit the txn.

**Comment by idsj** [comment]
@andreimatei  I'm not using the transaction actually, and after the statement, all the nodes got errors. I used to catch this situation when I drop a table with 1m+ records and I thought the reason lies in the table size, but now I meet the same error for table of only 10k+ size.

**Comment by vivekmenezes** [comment]
I believe this is the intended behavior right. The name is available for
reuse only after thr table has been fully deleted.

**Comment by andreimatei** [comment]
Right, but the table is truncated and "fully deleted" by the schema changer, which (when run through the "synchronous path") blocks the client executing the `DROP`.

**Comment by idsj** [comment]
@andreimatei 
create a new cockroach cluster with three node(192.168.181.91/92/93);Then do follow:

root@192.168.181.91:26257> create database test;
CREATE DATABASE
root@192.168.181.91:26257> set database = test;
SET

/*
create a table named test_serial and insert 30000 records using the node(192.168.181.93);
*/

root@192.168.181.91:26257> show tables;
+-------------+
|    Table    |
+-------------+
| test_serial |
+-------------+
root@192.168.181.91:26257> select count(*) from test_serial;
+----------+
| count(*) |
+----------+
|    30000 |
+----------+
root@192.168.181.91:26257> drop table test_serial;
DROP TABLE
root@192.168.181.91:26257> show tables;
+-------------+
|    Table    |
+-------------+
| test_serial |
+-------------+
root@192.168.181.91:26257> select count(*) from test_serial;
pq: table is being deleted
root@192.168.181.91:26257> create table test_serial(a serial primary key,b string(30),c bool);
pq: table "test_serial" already exists
root@192.168.181.91:26257> drop table test_serial;
pq: table "test_serial" has been deleted
root@192.168.181.91:26257> show tables;
+-------------+
|    Table    |
+-------------+
| test_serial |
+-------------+
root@192.168.181.91:26257>  


and the node(192.168.181.93) got error.

**Comment by andreimatei** [comment]
I couldn't manage to reproduce this. @idsj would you mind confirming you're running a recent build of cockroach? When you launch it with `./cockroach start`, there should be a line in the log (or console, if you specify `--logtostderr`) like
```
build:     beta-20160609-328-g775699c @ 2016/06/24 18:59:25 (go1.6.2)
```

Also, can you confirm that the `DROP` statement takes a noticeable time to execute (a few secs)?

Also also would you mind trying to reproduce it using a single node cluster, just to see if that changes anything?

Thanks!

**Comment by idsj** [comment]
@andreimatei 
i use the beta-20160616 version.The version message is as follow：
Build Tag:   beta-20160616
Build Time:  2016/06/16 16:51:45
Platform:    linux amd64
Go Version:  go1.6.2
C Compiler:  gcc 4.9.2

yes,it execute a few seconds.

Did you reproduce it using the cluster with VM? It did not happen in VM environment,i have done too. It happened in the physical machine environment.

**Comment by andreimatei** [comment]
OK, managed to repro using three nodes. Looking.

**Comment by andreimatei** [comment]
For an update here, a number of problems have been made apparent by this. There's an outright bug that would sometimes lead to this behavior that's being fixed in #7504.
More fundamentally, there's problems with humongous transactions, such as the one we use for a truncation. For example: https://github.com/cockroachdb/cockroach/issues/627

Ways to avoid a huge transaction for truncation are being discussed here: https://github.com/cockroachdb/cockroach/issues/7499

**Comment by vivekmenezes** [comment]
We should fix this by renaming the table before dropping it.
